src/Handler/GitHandler.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace LastCall\DownloadsPlugin\Handler;
+
+use Composer\Composer;
+use Composer\Config;
+use Composer\IO\IOInterface;
+use Composer\Util\Git;
+use Composer\Util\ProcessExecutor;
+use Composer\Util\Filesystem;
+use React\Promise\PromiseInterface;
+
+class GitHandler extends BaseHandler
+{
+    const TMP_PREFIX = '.composer-extra-tmp-';
+
+    public function createSubpackage()
+    {
+        $pkg = parent::createSubpackage();
+        $pkg->setDistType('git');
+        return $pkg;
+    }
+
+    public function getTrackingFile()
+    {
+        $file = basename($this->extraFile['id']) . '-' . md5($this->extraFile['id']) . '.json';
+        return
+            dirname($this->getTargetPath()) .
+            DIRECTORY_SEPARATOR . self::DOT_DIR .
+            DIRECTORY_SEPARATOR . $file;
+    }
+
+    /**
+     * @param Composer $composer
+     * @param IOInterface $io
+     */
+    public function download(Composer $composer, IOInterface $io) {
+        $urlAndVersion = $this->getSubpackage()->getDistUrl();
+        $config = $composer->getConfig();
+        $wd = $this->getTargetPath();
+        $process = new ProcessExecutor($io);
+        $cfs = new Filesystem();
+        $git = new Git($io, $config, $process, $cfs);
+        if (file_exists($wd)) {
```

Comment:
I haven't tested, but I see these two code-paths (based on `file_exists()`), and it makes me imagine the following use-case:

1. In top-level `composer.json`, add `extra.downloads` definition with a `git` URL.
2. Run `composer install`.
3. In top-level `composer.json`, edit `extra.downloads` and change the `git` commit/tag.
4. Run `composer install`.

I guess the idea is that step 2 makes a new folder, and step 4 updates that same folder.

But what if that folder doesn't have any `.git` data? For example:

1. In top-level `composer.json`, add `extra.downloads` definition with a `zip` URL.
2. Run `composer install`.
3. In top-level `composer.json`, edit `extra.downloads` definition to a `git` URL.
4. Run `composer install`.

At step 4, shouldn't we clear out the folder before doing any git operations?

(*IIRC, this only matters for `extra.downloads` in the top-level folder. In other contexts, there's a higher level `rm -rf`.*)